src/printer.js

```diff
@@ -2330,19 +2337,46 @@ function nodeStr(node, options) {
     shouldUseSingleQuote = true;
   }
 
-  const result = jsesc(str, {
-    quotes: shouldUseSingleQuote ? "single" : "double",
-    wrap: true
-  });
+  const raw = options.parser === 'flow' ? node.raw : node.extra.raw;
+  const isSingleQuote = raw.charAt(0) === "'";
 
-  // Workaround a bug in the Javascript version of the flow parser where
-  // astral unicode characters like \uD801\uDC28 are incorrectly parsed as
-  // a sequence of \uFFFD.
-  if (options.parser === "flow" && result.indexOf("\\uFFFD") !== -1) {
-    return node.raw;
+  if (isSingleQuote === shouldUseSingleQuote) {
+    return raw;
   }
 
-  return result;
+  const nextQuote = shouldUseSingleQuote ? "'" : '"';
+
+  // When we change the outer quotes, we need to change the number of
+  // backslash before. In order to know how many backslash to put,
+  // I manually done the first few examples and extracted a general rule
+  // out of it.
+  //
+  // shouldSwitch === true
+  //
+  //   3 -> 3 "\\\'" -> '\\\''
+  //   2 -> 3 "\\'"  -> '\\\''
+  //   1 -> 1 "\'"   -> '\''
+  //   0 -> 1 "\'"   -> '\''
+  //
+  //   Rule: n -> Math.floor(n / 2) + 1
+  //
+  // shouldSwitch === false
+  //
+  //   3 -> 2 '\\\'' -> "\\'"
+  //   1 -> 0 '\''   -> "'"
+  //
+  //   Rule: n -> n - 1
+  return nextQuote + raw.slice(1, -1)
+    .replace(/\\*['"]/g, (match) => {
+      const quote = match[match.length - 1];
+      const shouldSwitch = (quote === "'") === shouldUseSingleQuote;
+      const numBackslash = match.length - 1;
+      const nextNumBackslash = shouldSwitch ?
+        Math.floor(numBackslash / 2) + 1 :
+        numBackslash - 1;
```

Comment:
This looks pretty good. I'm still missing how this works if we are switching a single quoted string to a double quoted string, and we need to do the switching here. Is that handled?